framework/base/Controller.php

```diff
@@ -248,17 +248,29 @@ public function createAction($id)
 
         $actionMap = $this->actions();
         if (isset($actionMap[$id])) {
-            return Yii::createObject($actionMap[$id], [$id, $this]);
+            // external/standalone action
+            return $actionMap[$id] instanceof Action ? $actionMap[$id] : Yii::createObject($actionMap[$id], [$id, $this]);
         }
 
-        if (preg_match('/^(?:[a-z0-9_]+-)*[a-z0-9_]+$/', $id)) {
-            $methodName = 'action' . str_replace(' ', '', ucwords(str_replace('-', ' ', $id)));
-            if (method_exists($this, $methodName)) {
+        if (preg_match('/^(?:[a-z\d_]+-)*[a-z\d_]+$/', $id) === 1) {
+            $methodName = 'action' . str_replace('-', '', ucwords($id, '-'));
+            
+            $method = null;
+            if ($this->hasMethod($methodName, false)) {
                 $method = new \ReflectionMethod($this, $methodName);
-                if ($method->isPublic() && $method->getName() === $methodName) {
-                    return new InlineAction($id, $this, $methodName);
+            } else {
+                foreach ($this->getBehaviors() as $behavior) {
+                    if ($behavior->hasMethod($methodName)) {
+                        $method = new \ReflectionMethod($behavior, $methodName);
```

Comment:
The more behaviors there are attached the slower resolving will become even for apps that do not have actions defined via behaviors. It may not be noticeable though since behaviors are rarely used for controllers overall.

Reply:
@samdark it's plan C, if A (standalone action) and B (inline action defined in controller) not detected and runs not often.

Reply:
Well, it will slow down 404 at least :)